metric_learn/rca.py

```diff
@@ -62,32 +79,42 @@ def fit(self, data, chunks):
         when ``chunks[i] == j``, point i belongs to chunklet j.
     """
     data, chunks, num_chunks, d = self._process_inputs(data, chunks)
-
-    # mean center
-    data -= data.mean(axis=0)
-
-    # mean center each chunklet separately
-    chunk_mask = chunks != -1
-    chunk_data = data[chunk_mask]
-    chunk_labels = chunks[chunk_mask]
-    for c in xrange(num_chunks):
-      mask = chunk_labels == c
-      chunk_data[mask] -= chunk_data[mask].mean(axis=0)
-
-    # "inner" covariance of chunk deviations
+    chunk_mask, chunk_data, chunk_labels = _process_chunks(data, chunks, num_chunks)
     inner_cov = np.cov(chunk_data, rowvar=0, bias=1)
+    rank = np.linalg.matrix_rank(inner_cov)
+
+    # If the inner covariance matrix is not full rank,
+    # the input data are first projected with a PCA to a space of dimension rank.
+    M_pca = np.identity(d)
+    if rank < d:
+      pca = decomposition.PCA(n_components = rank)
+      pca.fit(data)
+      M_pca = pca.components_
+      data = pca.transform(data)
+      chunk_mask, chunk_data, chunk_labels = _process_chunks(data, chunks, num_chunks)
+      inner_cov = np.cov(chunk_data, rowvar=0, bias=1)
+      rank = np.linalg.matrix_rank(inner_cov)
```

Comment:
I have addressed all your remarks apart from this one. I have a few remarks/questions.

   The 3 papers presenting RCA (see below) emphasize that it is really important to perform a PCA before RCA to scale down noise dimensions. However, several methods are proposed to apply the PCA.

- Embedding dimension, denoted by `r`
   - In the paper ECCV 2002, `r` is set to the effective rank of the inner covariance matrix, i.e. the number of singular values that are significantly greater than 0 (see algorithm p.7).
   - Currently, I set `r` to the rank of the inner covariance matrix, i.e. the number of singular values that are not null. Perhaps, we should compute the effective rank as proposed in ECCV 2002. But how do we set the threshold `t` ? We should perhaps run some tests.
   - The paper JMLR 2005 presents another value for `r` computed from the size of the chunklets (see Algorithm 2 p.12). I do not really understand why this value is a good choice, and the computation depends on a parameter `alpha` (between 0 and 1). How do we set this parameter ?

- Rank of the inner covariance matrix after PCA
The PCA components are computed from the inner covariance matrix and are used to project the original data, and finally the inner covariance matrix is recomputed. The final inner covariance matrix is not the projection of the original covariance matrix on its PCA components.
Thus, as stated in the paper ECCV 2002,  the rank of the inner covariance matrix may become smaller than `r` after the projection. Perhaps, setting the threshold `t` high enough may prevent this issue. Another solution may be to keep projecting the original data to fewer principal components until the inner covariance matrix is full rank.

What I want to do :
  - Setting `r` to the the number of singular values of the inner covariance matrix that are significantly greater than 0. The remaining question is "how do I set the threshold" ?
  - Keeping on projecting the original data to fewer principal components until the inner covariance matrix is full rank. This way the implementation is robust, and we make no optimistic assumption about the rank of the inner covariance matrix.

What do you think of this proposal ?

The papers presenting RCA :
1. Adjustment learning and relevant component analysis, [ECCV 2002](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.19.2871&rep=rep1&type=pdf)
2. Learning distance functions using equivalence relations, [ICML 2003](https://pdfs.semanticscholar.org/0029/cf474bce70cfab2e944c4be01f99e741f1f4.pdf)
3. Learning a mahalanobis metric from equivalence constraints, [JMLR 2005](http://www.jmlr.org/papers/volume6/bar-hillel05a/bar-hillel05a.pdf)